From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Wed, 23 Nov 2022 22:03:33 +0100
Subject: [PATCH] Make EntityCollisionContext a live representation

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Make EntityCollisionContext a live representation"
By: Paul Sauve <paul@technove.co>
As part of: Airplane (https://github.com/TECHNOVE/Airplane)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

* Airplane description *

While Context is in the name, it is not used as a context. Instead it is
always created, use temporarily, then thrown away. This means having a
lot of fields to initialize and make space for is useless. I cannot find
anywhere in the codebase where this is used as a context which may be
saved for later, so this should be safe assuming plugins don't use it
for some strange reason.

* Airplane copyright *

Airplane
Copyright (C) 2020 Technove LLC

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

diff --git a/net/minecraft/world/phys/shapes/EntityCollisionContext.java b/net/minecraft/world/phys/shapes/EntityCollisionContext.java
index 98787a3e5366c4964db278785fbbd578373be991..24ef88bc86382ef3433116783e316aec65e9b646 100644
--- a/net/minecraft/world/phys/shapes/EntityCollisionContext.java
+++ b/net/minecraft/world/phys/shapes/EntityCollisionContext.java
@@ -12,16 +12,23 @@ import net.minecraft.world.level.material.FluidState;
 import org.jspecify.annotations.Nullable;
 
 public class EntityCollisionContext implements CollisionContext {
+    // Gale start - Airplane - make EntityCollisionContext a live representation - remove these and pray no plugin uses them
+    /*
     private final boolean descending;
     private final double entityBottom;
     private final boolean placement;
     private final ItemStack heldItem;
+    */
+    private final boolean placement;
+    // Gale end - Airplane - make EntityCollisionContext a live representation - remove these and pray no plugin uses them
     private final boolean alwaysCollideWithFluid;
     private final @Nullable Entity entity;
 
     protected EntityCollisionContext(
         boolean descending, boolean placement, double entityBottom, ItemStack heldItem, boolean alwaysCollideWithFluid, @Nullable Entity entity
     ) {
+        // Gale start - Airplane - make EntityCollisionContext a live representation - remove these and pray no plugin uses them
+        /*
         this.descending = descending;
         this.placement = placement;
         this.entityBottom = entityBottom;
@@ -40,11 +47,16 @@ public class EntityCollisionContext implements CollisionContext {
             alwaysCollideWithFluid,
             entity
         );
+        */
+        this.placement = placement;
+        this.alwaysCollideWithFluid = alwaysCollideWithFluid;
+        this.entity = entity;
+        // Gale end - Airplane - make EntityCollisionContext a live representation - remove unneeded things
     }
 
     @Override
     public boolean isHoldingItem(Item item) {
-        return this.heldItem.is(item);
+        return this.entity instanceof LivingEntity livingEntity ? livingEntity.getMainHandItem().is(item) : ItemStack.EMPTY.is(item); // Gale - Airplane - make EntityCollisionContext a live representation
     }
 
     @Override
@@ -64,12 +76,12 @@ public class EntityCollisionContext implements CollisionContext {
 
     @Override
     public boolean isDescending() {
-        return this.descending;
+        return this.entity != null && this.entity.isDescending(); // Gale - Airplane - make EntityCollisionContext a live representation
     }
 
     @Override
     public boolean isAbove(VoxelShape shape, BlockPos pos, boolean canAscend) {
-        return this.entityBottom > pos.getY() + shape.max(Direction.Axis.Y) - 1.0E-5F;
+        return (this.entity == null ? -Double.MAX_VALUE : entity.getY()) > (double) pos.getY() + shape.max(Direction.Axis.Y) - (double) 1.0E-5F; // Gale - Airplane - make EntityCollisionContext a live representation
     }
 
     public @Nullable Entity getEntity() {
